src/vopr/main.go

```diff
@@ -0,0 +1,766 @@
+	}
+}
+
+// Responsable for running the Vopr, capturing the output, processing it, writing it to file and
+// making the GitHub issue.
+// It also checks for duplicate issues and ensures the specified commit is available.
+func process(message vopr_message) {
+	commit_string := hex.EncodeToString(message.commit[:])
+
+	// Bugs 1 & 2 don't require a stack trace to be deduped
+	if dedupe_bug_1_and_2(message) == "duplicate" {
+		return
+	}
+
+	error := checkout_commit(commit_string, message.hash[:])
+	if error == nil {
+		log_debug("Successfully checked out commit "+commit_string, message.hash[:])
+	} else {
+		error_message := fmt.Sprintf(
+			"Failed to checkout commit %s: %s", commit_string,
+			error.Error(),
+		)
+		log_error(error_message, message.hash[:])
+		return
+	}
+
+	var output vopr_output
+	run_vopr(message.seed, &output, message.hash[:])
+
+	output.extract_stack_trace(&message)
+	output.parse_stack_trace(&message)
+	output.hash_stack_trace(&message)
+
+	issue_file_name := generate_file_name(message, output.stack_trace_hash)
+
+	// Bug 3 requires a stack trace to be deduped
+	if message.bug == 3 && dedupe(issue_file_name, message.hash[:]) == "duplicate" {
+		return
+	}
+
+	// Saves report to disk
+	create_issue_file(issue_file_name, &output, message.hash[:])
+
+	create_github_issue(message, &output, issue_file_name)
+}
+
+// Checks if a duplicate issue has already been submitted.
+func dedupe(issue_file_name string, message_hash []byte) string {
+	if _, error := os.Stat(issue_file_name); error == nil {
+		log_info("Duplicate issue found", message_hash)
+		return "duplicate"
+	} else {
+		return ""
+	}
+}
+
+// Bugs 1 and 2 don't require a stack trace for deduping and so they can be deduped before the Vopr
+// is run.
+func dedupe_bug_1_and_2(message vopr_message) string {
+	// If bug type 1 or 2 first check if file exists before parsing and hashing the stack trace
+	if message.bug == 1 || message.bug == 2 {
+		issue_file_name := generate_file_name(message, "")
+		return dedupe(issue_file_name, message.hash[:])
+	} else {
+		return ""
+	}
+}
+
+// Filenames are used to uniquely identify issues for deduping purposes.
+// correctness: 1_seed_commithash
+// liveness: 2_seed_commithash
+// crash: 3_commithash_stacktracehash
+func generate_file_name(message vopr_message, stack_trace_hash string) string {
+	commit_string := hex.EncodeToString(message.commit[:])
+
+	file_name := ""
+	if message.bug == 1 || message.bug == 2 {
+		file_name = fmt.Sprintf(
+			"%s/%d_%d_%s",
+			issue_directory,
+			message.bug,
+			message.seed,
+			commit_string,
+		)
+	} else if message.bug == 3 {
+		file_name = fmt.Sprintf(
+			"%s/%d_%s_%s",
+			issue_directory,
+			message.bug,
+			commit_string,
+			stack_trace_hash,
+		)
+	}
+	return file_name
+}
+
+// Fetch available branches from GitHub and checkout the correct commit if it exists.
+func checkout_commit(commit string, message_hash []byte) error {
```

Comment:
It would be good to double-check here at the last moment that the commit string is all hexadecimal. Otherwise, an attacker could inject other commands when we call git.

Reply:
Here's how these attacks work: https://owasp.org/www-community/attacks/Command_Injection